pkg/bootflow/examples/bruteforce/main.go

```diff
@@ -0,0 +1,131 @@
+package main
+
+import (
+	"bytes"
+	"crypto/sha1"
+	"fmt"
+	"hash"
+
+	"github.com/9elements/converged-security-suite/v2/pkg/bootflow/bootengine"
+	"github.com/9elements/converged-security-suite/v2/pkg/bootflow/steps/tpmsteps"
+	"github.com/9elements/converged-security-suite/v2/pkg/bootflow/trustchains/tpm"
+	"github.com/9elements/converged-security-suite/v2/pkg/bootflow/types"
+	"github.com/9elements/converged-security-suite/v2/pkg/bruteforcer"
+	"github.com/google/go-tpm/tpm2"
+)
+
+func main() {
+	// Here we brute-force a TPM measurement against a final PCR0 value.
+	//
+	// We initially have measurements of two separators, but we need to
+	// bruteforce the first measurement to 0x01000001
+
+	// an artificial flow for two separators
+	myFlow := types.Flow{
+		tpmsteps.InitTPMLazy(0),
+		tpmsteps.MeasureSeparator(0),
+		tpmsteps.MeasureSeparator(0),
+	}
+
+	// the expected final hash (after bruteforcing)
+	expectedHash := make([]byte, sha1.Size)
+	extend := func(h hash.Hash, a, b []byte) []byte {
+		defer h.Reset()
+		h.Write(a)
+		h.Write(b)
+		return h.Sum(nil)
+	}
+	sep0, sep1 := sha1.Sum([]byte("\x01\x00\x00\x01")), sha1.Sum([]byte("\x00\x00\x00\x00"))
+	expectedHash = extend(sha1.New(), expectedHash, sep0[:])
+	expectedHash = extend(sha1.New(), expectedHash, sep1[:])
+
+	// executing the flow (with two simple "\x00\x00\x00\x00" measurements)
+	state := types.NewState()
+	state.SetFlow(myFlow, 0)
+	state.EnableTrustChain(tpm.NewTPM())
+	process := bootengine.NewBootProcess(state)
+	process.Finish()
+
+	// just some debugging
+	fmt.Printf("Log:\n%#v\n", process.Log)
+
+	// == now let's bruteforce the first measurement ==
+
+	// first find the original (UNHASHED) data of the first measurement
+	tpmMeasurements := process.Log.GetVerifiedDataByTrustChainType((*tpm.TPM)(nil))
```

Comment:
design nit; this typed nil (although valid syntax) may look weird for less experienced. is there a way to express extensibility (external i suppose in this case) without using the type system like this? id also be interested in understanding more regarding the extensibility aspect (as i see it used a lot in this whole design) and why you'd want to encode it like this (rather than, for example, leave consumers to derive their own code or contrib to this codebase directly)

Reply:
> design nit; this typed nil (although valid syntax) may look weird for less experienced.

It looks weird even for myself :D

> is there a way to express extensibility (external i suppose in this case) without using the type system like this? 

Well. It does not require a `nil`. One may call for example:
```
tpmMeasurements := process.Log.GetVerifiedDataByTrustChainType(tpm.TPM{})
```
instead, and it still will work.

Or we can select not by type, but by a specific instance:
```
tpmInstance := tpm.NewTPM()
state.EnableTrustChain(tpmInstance)
[...]
tpmMeasurements := process.Log.GetVerifiedDataByTrustChain(tpmInstance)
```
would that be better?

Reply:
I think this is an artifact of the design. You're essentially using the type system as an enum value here (and discarding the actual instance value, which here is nil, or even worse tpm.TPM{}). It's usually more clear to be direct about it (as in, use an actual enum), but then again, I don't have a simple localized proposal for this right now considering all the other code

Reply:
Changed to:
```
tpmMeasurements := process.Log.GetDataVerifiedBy(tpmInstance)
```